File: python/src/problem/leetcode/medium/leetcode_2.py
# coding=utf-8

from src.data_structure.data_structure import ListNode

class Solution:

    def add_two_numbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        """
        Time: O(n), Space: O(1)
        :param l1:
        :param l2:
        :return:
        """
        p = dummy = ListNode(-1)
        carry = 0
        while l1 or l2 or carry > 0:
            sum = carry
            if l1:
                sum += l1.val
                l1 = l1.next
            if l2:
                sum += l2.val
                l2 = l2.next
            p.next = ListNode(sum % 10)
            p = p.next
            carry = sum // 10
        return dummy.next

File: c/medium/leetcode2.c
//
// Created by ywh on 2020/7/20.
//

#include <stdlib.h>
#include "../dataStructure.h"

/**
 * 求两个单链表之和
 *
 * @param l1
 * @param l2
 * @return
 */
struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2) {
    int carry = 0, sum;
    struct ListNode dummy, *cur = l1, *prev = &dummy;
    dummy.next = cur;
    while (l1 || l2) {
        sum = 0;
        if (l1) {
            sum += l1->val;
            l1 = l1->next;
        }
        if (l2) {
            if (!cur) {
                prev->next = l2;
                cur = l2;
            }
            sum += l2->val;
            l2 = l2->next;
        }
        sum += carry;
        carry = sum / 10;
        cur->val = sum % 10;
        prev = cur;
        cur = cur->next;
    }
    if (carry) {
        cur = malloc(sizeof(*cur));
        cur->val = carry;
        cur->next = NULL;
        prev->next = cur;
    }
    return dummy.next;
}

/**
 * 求两个单链表之和
 *
 * @param l1
 * @param l2
 * @return
 */
struct ListNode *addTwoNumbers2(struct ListNode *l1, struct ListNode *l2) {
    int carry = 0, sum;
    struct ListNode dummy, *p = &dummy, *newNode;
    while (l1 || l2 || carry > 0) {
        sum = 0;
        if (l1) {
            sum += l1->val;
            l1 = l1->next;
        }
        if (l2) {
            sum += l2->val;
            l2 = l2->next;
        }
        sum += carry;
        carry = sum / 10;
        newNode = (struct ListNode *) malloc(sizeof(struct ListNode));
        newNode->val = sum % 10;
        newNode->next = NULL;
        p->next = newNode;
        p = p->next;
    }
    return dummy.next;
}